lib/block-supports/background.php

```diff
@@ -30,6 +30,30 @@ function gutenberg_register_background_support( $block_type ) {
 	}
 }
 
+/**
+ * Given a theme.json or block background styles, returns the background styles for a block.
+ *
+ * @since 6.6.0
+ *
+ * @param  array $background_styles Background style properties.
+ * @return array                     Style engine array of CSS string and style declarations.
+ */
+function gutenberg_get_background_support_styles( $background_styles = array() ) {
+	$background_image_source = ! empty( $background_styles['backgroundImage']['source'] ) ? $background_styles['backgroundImage']['source'] : null;
+
+	/*
+	 * "theme" source implies relative path to the theme directory
+	 */
+	if ( ! empty( $background_styles['backgroundImage']['url'] ) && is_string( $background_styles['backgroundImage']['url'] ) && 'theme' === $background_image_source ) {
+		if ( ! str_starts_with( $background_styles['backgroundImage']['url'], '/' ) ) {
+			$background_styles['backgroundImage']['url'] = '/' . $background_styles['backgroundImage']['url'];
+		}
+		$background_styles['backgroundImage']['url'] = esc_url( get_stylesheet_directory_uri() . $background_styles['backgroundImage']['url'] );
```

Comment:
I think we could safely use [get_theme_file_uri](https://developer.wordpress.org/reference/functions/get_theme_file_uri/) here. It'll look in the active theme dir first (e.g., active child theme), and then look to the parent theme directory if not found.

Why? In the case where a child theme is active, I can imagine that some themes might want to dictate that relative image paths should be prefixed with a theme parent directory URI, and not the active child theme's.

This could also be expressed explicitly in theme.json later if we come up with a decent structure/naming

```json
	"styles": {
		"background": {
			"backgroundImage": {
				"source": "theme-child", // or "theme-parent" if active is default, or "theme-inherit" or 🤷🏻 
				"url": "img/Untitled.jpg"
			},
		}
	}
```

Reply:
The above comment doesn't consider the editor or background images in template or pattern files 🤔 

So perhaps we go the theme.json explicit route. It would be more flexible for theme developers anyway, in the sense that they could mix and match where images came from, e.g., site background image from the parent theme and other from child theme.

Reply:
> I think we could safely use [get_theme_file_uri](https://developer.wordpress.org/reference/functions/get_theme_file_uri/) here.

Good catch! Yes, that logic looks good for allowing child themes to override an image from a parent theme, with fallback to the parent theme when the child theme doesn't override it 👍

> So perhaps we go the theme.json explicit route. It would be more flexible for theme developers anyway, in the sense that they could mix and match where images came from, e.g., site background image from the parent theme and other from child theme.

That sounds like a good potential idea for a follow-up to me. Allowing child themes to universally override something from a parent theme conforms to the general principle of child theming, and I'm curious if many folks would intentionally want to go with the exception? (Explicitly using `theme-parent` to skip an overridden image in a child theme feels potentially a bit edge-casey to me).

What blockers would we encounter if we go with `"source": "theme"` for now and attempt to have the fallback logic be: use the child theme asset if available, and fallback to the parent theme if not?

Reply:
Thanks for chiming in! I'm probably overthinking it. 

I'm only raising this now as this PR introduces `"source": "theme"` so I wanted to consider this angle while we have the chance to redefine it.

> What blockers would we encounter if we go with "source": "theme" for now and attempt to have the fallback logic be: use the child theme asset if available, and fallback to the parent theme if not?

This would be ideal, and I think the way forward. Especially seeing a child theme inherits templates, theme. json etc from the parent theme.

The only splinter is reproducing the file checking in the editor, which could be a performance hit assuming we set up some API. 

🤔

Reply:
> The only splinter is reproducing the file checking in the editor, which could be a performance hit assuming we set up some API.

Unless we come up with a way to reliably parse template/pattern files.

This PR as it is should be okay just for theme.json, but I'm concerned that if a child theme is using its parents' templates/patterns the file paths need to be resolved with the parent's URI and not the child's.

It's all very spicy!

Reply:
> I'm only raising this now as this PR introduces "source": "theme" so I wanted to consider this angle while we have the chance to redefine it.

Good idea, this is the right time to be thinking through all the options, for sure 👍

> The only splinter is reproducing the file checking in the editor, which could be a performance hit assuming we set up some API.

Yeah, the lightest weight version I could think of would be to do a `HEAD` request using the child theme url, and if we get a 404, then use the parent them url, or something like that?

Reply:
> Yeah, the lightest weight version I could think of would be to do a HEAD request using the child theme url, and if we get a 404, then use the parent them url, or something like that?

I've been investigating this, and there's a slight issue in that it's an async call that needs to be made in a series of synchronous steps, either when parsing the styles or the block props. 

I'm a bit stumped as to where the relative URL parsing should therefore take place without impacting performance.


For now, in the editor, I'm testing doubling up the images, that is, adding both urls to the `background-image` property with priority given to the `stylesheetURI`, e.g., 


```css
background-image: url("http://localhost:8888/wp-content/themes/emptytheme-child/img/untitled.png"), url("http://localhost:8888/wp-content/themes/emptytheme/img/untitled.png");
```

Where the first image doesn't load, the second image is displayed.

This works of course, and can even be combined with linear gradients, 

```css
background-image: linear-gradient(rgba(0, 0, 255, 0.5), rgba(255, 255, 0, 0.5)), url("http://localhost:8888/wp-content/themes/emptytheme-child/img/untitled.png"), url("http://localhost:8888/wp-content/themes/emptytheme/img/untitled.png");
```


but I hope it doesn't bite us in the derrière later when we deal with more complicated layers. I can't see how yet though. Even if we implement layers as described in https://github.com/WordPress/gutenberg/pull/60739#issuecomment-2081872934, we could duplicate the properties for each image layer in the editor. 🤷🏻 

Anyway, that's enough for me today.

Reply:
> For now, in the editor, I'm testing doubling up the images, that is, adding both urls to the background-image property with priority given to the stylesheetURI, e.g.

Oh, that's pretty clever! The main edge case I can think of is if a theme includes a transparent image like a parchment texture that's meant to be combined with a background color, and if a child theme wanted to override that with a different transparent texture. If we add the two together, it sounds like that would then layer the two images, instead of visually overriding them?

Reply:
> The main edge case I can think of is if a theme includes a transparent image like a parchment texture that's meant to be combined with a background color, and if a child theme wanted to override that with a different transparent texture. If we add the two together, it sounds like that would then layer the two images, instead of visually overriding them?

Yeah, that sounds like a risk. Thanks for raising it. 

I'm tending towards accepting it as a reasonable risk in order to get the feature working, at least while we come up with a better approach.

I'm also going to keep exploring - something might drop into my lap 😅